### PR TITLE
refactor settings, add AwardYear scope

### PR DESCRIPTION
**Pull request by dreamfall**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -124,4 +124,15 @@ def need_authentication?
     (Rails.env.staging? || Rails.env.production?) &&
     controller_name != "healthchecks"
   end
+
+  def load_award_year_and_settings
```
Comment (matthewford) [inline app/controllers/application_controller.rb]: Assignment Branch Condition size for load_award_year_and_settings is too high. [15.65/15]

File: app/models/form_answer.rb [hunk 2]
```diff
@@ -182,7 +182,7 @@ def set_urn
 
     next_seq = self.class.connection.select_value("SELECT nextval(#{ActiveRecord::Base.sanitize("urn_seq_#{award_type}")})")
 
-    self.urn = "QA#{sprintf('%.4d', next_seq)}/#{award_year.to_s[2..-1]}#{award_type[0].capitalize}"
+    self.urn = "QA#{sprintf('%.4d', next_seq)}/#{award_year.year.to_s[2..-1]}#{award_type[0].capitalize}"
```
Comment (matthewford) [inline app/models/form_answer.rb]: Line is too long. [105/100]

File: app/models/form_answer.rb [hunk 3]
```diff
@@ -38,6 +37,7 @@ class FormAnswer < ActiveRecord::Base
   begin :associations
     belongs_to :user
     belongs_to :account
+    belongs_to :award_year
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 4]
```diff
@@ -90,15 +90,15 @@ def secondary
   end
 
   begin :scopes
-    scope :for_award_type, -> (award_type) { where award_type: award_type }
-    scope :for_year, -> (year) { where award_year: year }
+    scope :for_award_type, -> (award_type) { where(award_type: award_type) }
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 5]
```diff
@@ -90,15 +90,15 @@ def secondary
...
+    scope :for_year, -> (year) { joins(:award_year).where(award_years: { year: year }) }
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/models/form_answer.rb [hunk 6]
```diff
@@ -90,15 +90,15 @@ def secondary
...
     scope :shortlisted_with_no_certificate, -> { where("1 = 0") }
     scope :winners, -> { where("1 = 0") }
     scope :unsuccessful, -> { where(state: %w(not_recommended reserved)) }
   end
 
   begin :callbacks
-    before_save :set_award_year, unless: :award_year?
+    before_save :set_award_year, unless: :award_year
```
Comment (matthewford) [inline app/models/form_answer.rb]: Inconsistent indentation detected.

File: app/services/notifiers/email_notification_service.rb [hunk 7]
```diff
@@ -24,32 +25,32 @@ def run
     end
   end
 
-  def shortlisted_notifier
+  def shortlisted_notifier(award_year)
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Unused method argument - `award_year`. If it's necessary, use `_` or `_award_year` as an argument name to indicate that it won't be used. You can also write as `shortlisted_notifier(*)` if you want the method to accept any arguments but don't care about them.

File: app/services/notifiers/email_notification_service.rb [hunk 8]
```diff
@@ -24,32 +25,32 @@ def run
...
     User.shortlisted.each do |user|
       Users::NotifyShortlistedMailer.notify(user).deliver_later!
     end
   end
 
-  def not_shortlisted_notifier
+  def not_shortlisted_notifier(award_year)
```
Comment (matthewford) [inline app/services/notifiers/email_notification_service.rb]: Unused method argument - `award_year`. If it's necessary, use `_` or `_award_year` as an argument name to indicate that it won't be used. You can also write as `not_shortlisted_notifier(*)` if you want the method to accept any arguments but don't care about them.

File: spec/support/shared_contexts/admin_feedback_pdf_file_checks.rb [hunk 9]
```diff
@@ -44,7 +44,7 @@
   end
 
   let(:award_general_information) do
-    "#{FeedbackPdfs::General::DrawElements::AWARD_GENERAL_INFO_PREFIX} #{form_answer.award_year}"
+    "#{FeedbackPdfs::General::DrawElements::AWARD_GENERAL_INFO_PREFIX} #{form_answer.award_year.year}"
```
Comment (matthewford) [inline spec/support/shared_contexts/admin_feedback_pdf_file_checks.rb]: Line is too long. [102/100]

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 10]
```diff
@@ -3,22 +3,22 @@
 describe "FeedbackPdfs::Base" do
   let!(:form_answer_2014_innovation) do
     FactoryGirl.create :form_answer, :submitted, :innovation,
-                        award_year: 2014
+                        award_year: AwardYear.where(year: 2014).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 11]
```diff
@@ -3,22 +3,22 @@
...
   end
 
   let!(:form_answer_2015_innovation) do
     FactoryGirl.create :form_answer, :submitted, :innovation,
-                        award_year: 2015
+                        award_year: AwardYear.where(year: 2015).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 12]
```diff
@@ -3,22 +3,22 @@
...
   end
 
   let!(:form_answer_2014_trade) do
     FactoryGirl.create :form_answer, :submitted, :trade,
-                        award_year: 2014
+                        award_year: AwardYear.where(year: 2014).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: spec/unit/pdf_generators/feedback_pdfs/base_spec.rb [hunk 13]
```diff
@@ -3,22 +3,22 @@
...
   end
 
   let!(:form_answer_2015_trade) do
     FactoryGirl.create :form_answer, :submitted, :trade,
-                        award_year: 2015
+                        award_year: AwardYear.where(year: 2015).first_or_create
```
Comment (matthewford) [inline spec/unit/pdf_generators/feedback_pdfs/base_spec.rb]: Align the parameters of a method call if they span more than one line.

File: app/services/notifiers/email_notification_service.rb (hunk 7 above)
Comment (dreamfall) [inline app/services/notifiers/email_notification_service.rb]: I will be used later.

**Comment by jaleszek** [comment]
reviewing

**Comment by dreamfall** [comment]
@jaleszek I just thought it's worth adding one test for email notifications.

File: app/models/award_year.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  has_many :form_answers
+  has_one :settings, inverse_of: :award_year, autosave: true
+
+  after_create :create_settings
+
+  AVAILABLE_YEARS = [2016, 2017, 2018, 2019]
+
+  # +1 here is to match URN number of assosiated forms
+  def self.current
+    for_year(Date.current.year + 1).first_or_create
```
Comment (jaleszek) [inline app/models/award_year.rb]: I though we don't want to have application awarding year associated with calendar year ? How it works when 2 awarding campaigns happens during one year or one exceeds the New Year and covers 2 calendar years ?

**Pull request closed without merging**